### PR TITLE
GAWB-3138: Attempt to fix a flakytest

### PR DESCRIPTION
**Pull request by jmthibault79**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by jmthibault79** [comment]
This test was often failing with:
```
[info] - should grant and remove google role access *** FAILED ***
[info]   The future returned an exception of type: java.util.NoSuchElementException, with message: Future.failed not completed with a throwable.. (OrchestrationApiSpec.scala:62)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.concurrent.Futures$FutureConcept$class.tryTryAgain$1(Futures.scala:523)
[info]   at org.scalatest.concurrent.Futures$FutureConcept$class.futureValueImpl(Futures.scala:550)
[info]   at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValueImpl(ScalaFutures.scala:275)
[info]   at org.scalatest.concurrent.Futures$FutureConcept$class.futureValue(Futures.scala:476)
[info]   at org.scalatest.concurrent.ScalaFutures$$anon$1.futureValue(ScalaFutures.scala:275)
[info]   at org.broadinstitute.dsde.firecloud.test.api.orch.OrchestrationApiSpec$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$3.apply(OrchestrationApiSpec.scala:62)
```

For a better error message, I updated a `Retry` to an `eventually` which is natively test-aware.

For more reliable tests, I increased the possible wait time from 10 sec to 1 minute, and added an additional instance of `eventually` to cover both google role transition times.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala [hunk 1]
```diff
@@ -51,7 +50,10 @@ class OrchestrationApiSpec extends FreeSpec with Matchers with ScalaFutures with
 
         Orchestration.billing.addGoogleRoleToBillingProjectUser(projectName, user.email, role)(ownerToken)
 
-        val queryReference = googleBigQueryDAO.startQuery(userToken.value, GoogleProject(projectName), shakespeareQuery).futureValue
+        val queryReference = eventually {
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala]: I'm not sure it makes sense to use both `eventually` and `.futureValue`? I think calling `.futureValue` already takes the `PatienceConfig` into account when polling for a result.

Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala]: hm ok - I'll try without

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala [hunk 2]
```diff
@@ -9,20 +9,19 @@ import org.broadinstitute.dsde.workbench.dao.Google.googleBigQueryDAO
...
-import scala.util.Try
 
 class OrchestrationApiSpec extends FreeSpec with Matchers with ScalaFutures with Eventually
   with BillingFixtures {
   implicit override val patienceConfig: PatienceConfig = PatienceConfig(timeout = scaled(Span(5, Seconds)))
 
   "Orchestration" - {
     "should grant and remove google role access" in {
+      // google roles may take a little while to take effect
+      implicit val patienceConfig: PatienceConfig = PatienceConfig(timeout = scaled(Span(1, Minute)))
```
Comment (ansingh7115) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala]: default value for interval is 15 milliseconds I believe - is that too often or are we good with that?

Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala]: That seems excessive.  I can increase.  Say 1 second?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala (hunk 1 above)
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala]: facetalked - modified my approach, did not remove

**Review by rtitle — APPROVED** [review]
The code looks good, hope it fixes flakiness

**Review by ansingh7115 — APPROVED** [review]
Yay! All hail Joel, slayer of flakey tests! :)

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala [hunk 3]
```diff
@@ -44,14 +43,22 @@ class OrchestrationApiSpec extends FreeSpec with Matchers with ScalaFutures with
...
         preRoleFailure.getMessage should include(projectName)
         preRoleFailure.getMessage should include("bigquery.jobs.create")
 
         Orchestration.billing.addGoogleRoleToBillingProjectUser(projectName, user.email, role)(ownerToken)
 
-        val queryReference = googleBigQueryDAO.startQuery(userToken.value, GoogleProject(projectName), shakespeareQuery).futureValue
+        // The google role might not have been applied the first time we call startQuery() - poll until it has
+        val queryReference = eventually {
+          val start = googleBigQueryDAO.startQuery(userToken.value, GoogleProject(projectName), shakespeareQuery).futureValue
+          start shouldBe a[JobReference]
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala:57]: I think I understand this case... you're retrying the thing until the thing returns the type of result you expect if the thing was successful.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala [hunk 4]
```diff
@@ -63,15 +70,16 @@ class OrchestrationApiSpec extends FreeSpec with Matchers with ScalaFutures with
...
 
-        val postRoleFailure = Retry.retry(1.second, 10.seconds) {
-          // retry this because removing google roles is not always immediate
-          Try { googleBigQueryDAO.startQuery(userToken.value, GoogleProject(projectName), shakespeareQuery).failed.futureValue }.toOption
-        }.get
+        // The google role might not have been removed the first time we call startQuery() - poll until it has
+        eventually {
+          val postRoleFailure = googleBigQueryDAO.startQuery(userToken.value, GoogleProject(projectName), shakespeareQuery).failed.futureValue
+
+          postRoleFailure shouldBe a[GoogleJsonResponseException]
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala]: If my understanding of the pattern above is correct, then I think the `should`s below this line should be outside of the `eventually`. You want to break out of the `eventually` as soon as you have the type of response you expect. By making the other checks inside, you'll also retry if any of the assertions fail. For example, if Google changes the message to not include the project name, the conditions will never be met and retries will happen unnecessarily.

**Review by breilly2 — CHANGES_REQUESTED** [review]
I'm going to boldly say "Request changes" based on my understanding of how this works, but please let me know if I'm wrong and I'll update my review.

**Comment by breilly2** [comment]
Thank you @jmthibault79 and all of the reviewers for putting the time and energy into improving this!

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala (hunk 4 above)
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/orch/OrchestrationApiSpec.scala]: OK, this change makes sense.

**Comment by jmthibault79** [comment]
Just noticed a copy/paste bug in the original after Brian pointed out an issue nearby.  This probably didn't cause flakiness, but may have caused it to pass incorrectly!

**Review by aweng98 — APPROVED** [review]

**Review by breilly2 — APPROVED** [review]

**Comment by breilly2** [comment]
Yikes! Glad you caught that!

**Comment by jmthibault79** [comment]
jenkins retest

**Pull request merged**